ui/src/app/shared/url-search-params.utils.ts

```diff
@@ -0,0 +1,103 @@
+import {QueryFields} from "./common";
+
+/** Utilities for working with URLSearchParams*/
+export class URLSearchParamsUtils {
+
+  /** Accepts a QueryJobsRequest, and translates it reversibly into a string that can be passed as a QueryParam. */
+  public static encodeURLSearchParams(request: QueryJobsRequest): string {
+    let urlSearchParams = new URLSearchParams();
+    urlSearchParams.set(QueryFields.parentId, request.parentId);
+    urlSearchParams.set(QueryFields.jobName, request.name);
+    for (let s in request.statuses) {
+      urlSearchParams.append('statuses', JobStatus[request.statuses[s]]);
+    }
+    if (request.start) {
+      urlSearchParams.set(QueryFields.start, request.start.toISOString());
+    }
+    if (request.end) {
+      urlSearchParams.set(QueryFields.end, request.end.toISOString());
+    }
+
+    if (request.labels) {
+      (request.labels as Map<String, String>).forEach((value: string, key: string) => {
+        urlSearchParams.set(key, value);
+      });
+    }
+    return urlSearchParams.toString();
+  }
+
+  /** Accepts a param map, and translates it reversibly into a string that can be passed as a QueryParam. */
+  public static encodeURLSearchParamsFromMap(params: Map<String, String[]>): string {
+    let urlSearchParams = new URLSearchParams();
+    params.forEach((values: string[], key: string) => {
+      if (values.length == 1) {
+        urlSearchParams.set(key, values[0]);
+      } else {
+        values.forEach(value => urlSearchParams.append(key, value));
+      }
+    });
+    return urlSearchParams.toString();
+  }
+
+  /** Accepts a string query that was generated via encodeUrlSearchParams and converts it back into a QueryJobsRequest. */
+  public static unpackURLSearchParams(query: string): QueryJobsRequest {
+    let urlSearchParams = new URLSearchParams(query);
+    let queryRequest = {
+      labels: new Map(),
+    };
+
+    urlSearchParams.paramsMap.forEach((values: string[], key: string) => {
+      if (key == QueryFields.parentId) {
+        queryRequest['parentId'] = urlSearchParams.get(QueryFields.parentId);
+      }
+      else if (key == QueryFields.jobName) {
+        queryRequest['name'] = urlSearchParams.get(QueryFields.jobName);
+      }
+      else if (key == QueryFields.statuses) {
+        let statuses: JobStatus[] = [];
+        for (let status of urlSearchParams.getAll(QueryFields.statuses)) {
+          statuses.push(JobStatus[status]);
+        }
+        queryRequest['statuses'] = statuses;
+      }
+      else if (key == QueryFields.start) {
+        queryRequest['start'] = new Date(urlSearchParams.get(QueryFields.start));
+      }
+      else if (key == QueryFields.end) {
+        queryRequest['end'] = new Date(urlSearchParams.get(QueryFields.end));
+      }
+      // Assume that any filters not matching a primary field should be interpreted as a label
+      else {
+        queryRequest.labels.set(key, urlSearchParams.get(key));
```

Comment:
This isn't getting serialized into the request properly. For example setting a `user-id` in the query field and clicking search doesn't change the request payload to the backend. Maybe its because this is a map rather than key-value?

Reply:
The dsub shim does have support for querying by labels and I'm pretty sure its working correctly––all the [tests](https://github.com/DataBiosphere/job-manager/blob/master/servers/dsub/jobs/test/base_test_cases.py#L288) are passing. Could you give a little bit more info about the 500 error you were seeing?

Which labels have you tried querying by, only `user-id` `job-id` and other `dsub` default ones ? `['task-id', 'job-id', 'user-id', 'job-name']` are reserved labels in dsub, so if I had to guess it is probably because we're trying to query them like normal labels, when we need to have a specific handling. A label like 'status-detail' should work, however.

I don't think this PR should fix the issue in the backend, but you should document the bug so that it can be fixed and the code which serializes labels into the request should function. Using the `Map` definitely will not work since its not even serialized into the request to the shim.

Reply:
I'm not familiar with the dsub repo and github tools for searching through code are non-existent so I haven't yet found the source of this error, but here's all the debugging information I have:

detail:"The server encountered an internal error and was unable to complete your request.  Either the server is overloaded or there is an error in the application."
status:500
title:"Internal Server Error"
type:"about:blank"

I can confirm this happens with status-detail as well as the reserved labels.

Reply:
Okay would you mind filing a github issue and adding a TODO wherever you disabled the non-supported fields. I'll try to figure out the issue and re-enable in another PR.